movebucket_test.go

```diff
@@ -0,0 +1,482 @@
+		isSrcKey      bool
+		isDstKey      bool
+		expActErr     error
+	}{
+		{
+			name:          "happy path",
+			srcBucketPath: []string{"rootBucketX", "parentBucketY", "myBucketZ"},
+			dstBucketPath: []string{"rootBucketA", "parentBucketB"},
+			bucketToMove:  "myBucketZ",
+			isSrcKey:      false,
+			isDstKey:      false,
+			expActErr:     nil,
+		},
+		{
+			name:          "bucketToMove not exist in srcBucket",
+			srcBucketPath: []string{"rootBucketX", "parentBucketY"},
+			dstBucketPath: []string{"rootBucketA", "parentBucketB"},
+			bucketToMove:  "myBucketZ",
+			isSrcKey:      false,
+			isDstKey:      false,
+			expActErr:     errors.ErrBucketNotFound,
+		},
+		{
+			name:          "bucketToMove exist in dstBucket",
+			srcBucketPath: []string{"rootBucketX", "parentBucketY", "myBucketZ"},
+			dstBucketPath: []string{"rootBucketA", "parentBucketB", "myBucketZ"},
+			bucketToMove:  "myBucketZ",
+			isSrcKey:      false,
+			isDstKey:      false,
+			expActErr:     errors.ErrBucketExists,
+		},
+		{
+			name:          "bucketToMove key exist in srcBucket, but no subBucket value",
+			srcBucketPath: []string{"rootBucketX", "parentBucketY"},
+			dstBucketPath: []string{"rootBucketA", "parentBucketB"},
+			bucketToMove:  "myBucketZ",
+			isSrcKey:      true,
+			isDstKey:      false,
+			expActErr:     errors.ErrIncompatibleValue,
+		},
+		{
+			name:          "bucketToMove key exist in dstBucket, but no subBucket value",
+			srcBucketPath: []string{"rootBucketX", "parentBucketY", "myBucketZ"},
+			dstBucketPath: []string{"rootBucketA", "parentBucketB"},
+			bucketToMove:  "myBucketZ",
+			isSrcKey:      false,
+			isDstKey:      true,
+			expActErr:     errors.ErrIncompatibleValue,
+		},
+		{
+			name:          "srcBucket is rootBucket",
+			srcBucketPath: []string{"", "myBucketZ"},
+			dstBucketPath: []string{"rootBucketA", "parentBucketB"},
+			bucketToMove:  "myBucketZ",
+			isSrcKey:      false,
+			isDstKey:      false,
+			expActErr:     nil,
+		},
+		{
+			name:          "dstBucket is rootBucket",
+			srcBucketPath: []string{"rootBucketX", "parentBucketY", "myBucketZ"},
+			dstBucketPath: []string{""},
+			bucketToMove:  "myBucketZ",
+			isSrcKey:      false,
+			isDstKey:      false,
+			expActErr:     nil,
+		},
+		{
+			name:          "srcBucket is rootBucket, and dstBucket is rootBucket",
+			srcBucketPath: []string{"", "myBucketZ"},
+			dstBucketPath: []string{""},
+			bucketToMove:  "myBucketZ",
+			isSrcKey:      false,
+			isDstKey:      false,
+			expActErr:     errors.ErrSameBuckets,
+		},
+	}
+
+	for _, tc := range testCases {
+
+		t.Run(tc.name, func(*testing.T) {
+			db := btesting.MustCreateDB(t)
+
+			dumpBucketBeforeMoving := tempfile()
+			dumpBucketAfterMoving := tempfile()
+
+			// arrange
+			if err := db.Update(func(tx *bbolt.Tx) error {
+				srcBucket := createBucketIfNotExist(t, tx, tc.srcBucketPath...)
+				if tc.isSrcKey {
+					pErr := srcBucket.Put([]byte(tc.bucketToMove), []byte("0"))
```

Comment:
It seems the key is included in the `bucketToMove`.
Shouldn't the key be in the same level as the `bucketToMove`?

Reply:
so this is covering the case where the name of the bucket to move is actually a key/value pair, not key/sub-bucket pair

so I insert the bucket name (i.e. bucketToMove) as a key / value pair in the src bucket 

please correct me if i am wrong

Reply:
Let's work with an example, assuming the db has the following hierarchical structure,
```
root 
    sb1
      sb2 
        sb3ToMove
    db1
      db2
```

and we are moving the `sb3ToMove`, and eventually the structure should be something like below,
```
root 
    sb1
      sb2 
    db1
      db2
        sb3ToMove
```

Now we plan to verify the case that an incompatible key exists in the source bucket. In this case, the `sb3ToMove` is actually an incompatible key instead of a bucket.

Your implementation isn't that clear to me. Proposed Pseudocode or steps for generating the sample data，
- Create hierarchical structure
```
root 
    sb1
      sb2 
    db1
      db2
```
- If `hasIncompatibleKeyInSrc` is `true`,  create an incompatible key `sb3ToMove` under `sb2`; otherwise, create a bucket `sb3ToMove` under `sb2`.
- If `hasIncompatibleKeyInDst` is `true`, create an incompatible key `sb3ToMove` under `db2`

I believe the step 2 (move the bucket) and step 3 (verify the result) should can be simplified as well after you make the proposed change above.

Reply:
done 👍🏽